### DISCUSSION TITLE
Announcing pyLoad v0.4.20

### DISCUSSION THREAD

**Discussion by GammaC0de** [pyload/pyload#3976] [Announcements]
7 Years, 4485 commits,
02/20/2020 - It time for pyLoad v0.4.20 🎉🎉🎉

https://github.com/pyload/pyload/releases/tag/0.4.20

### How to update?

Either:
- Install a realase package from https://github.com/pyload/pyload/releases/tag/0.4.20

OR:

- Go to `http://<PYLOAD_IP>/info` and write down the value of `Installation Folder` and `Config Folder`.
- Stop pyLoad.
- Delete the `<Config Folder>\userplugins` directory.
- Download the latest `stable` branch from https://github.com/pyload/pyload/archive/stable.zip
- Extract stable.zip into `<Installation Folder>` overwriting the old files (the zip file contains a `pyload-stable` root directory - you must extract the files without this directory, e.g: `pyload-stable\pyLoadCore.py` becomes `pyLoadCore.py` and so on..).
- Start pyLoad.

Good luck.

**Comment by StefftheEmperor2** [comment]
Hi there .. what do you think about working on the 0.5 release. I've got a working fork. But the versions of the plugins will diverge when we're working on different branches.

**Comment by GammaC0de** [comment]
@StefftheEmperor2 
> what do you think about working on the 0.5 release

I really want to get it happen
>  I've got a working fork

Where?

**Comment by morefreedom** [comment]
Whats the best was to update pyload whithout loosing any settings?
Can you please give some detailed instructions.

New issue: https://github.com/pyload/pyload/issues/3524

**Comment by LiloBzH** [comment]
Update ok . No problem at this time.

Thank you :)

**Comment by treysis** [comment]
"Error importing <package>: No module named wintypes"

Is this a problem with pyLoadCore.exe missing a python-package?

First line also stating: "INFO Starting pyLoad **0.4.9**"

**Comment by treysis** [comment]
pyLoadCore.exe will it be updated to reflect version 0.4.20?

**Comment by GammaC0de** [comment]
@treysis Are you running the windows version of pyload?

Just install python 2.7 and all the dependencies and you're done.

Using pyLoadCore.exe (which was made using py2exe) limits you from updating the core modules just by copying the .py files. so it is not recommended.

**Comment by treysis** [comment]
Yes, it's working with pyhton 2.7, however, with this I get some errors related to the outdated Crypto-library. I had to additionally install pyCrypto and replace the "Crypto" folder in the pyload "dist" directory with the new one.

Furthermore, I needed to place "wintypes.pyc" into "ctypes" folder for the following plugins to load: `UpdateManager, EventMapper, ExternalScripts, UnSkipOnFail, ExtractArchive, XFileSharing, ClickNLoad, UserAgentSwitcher`
I think this one is related to the new "import ctypes" from "misc.py".

I understand it's better to use local installation of python interpreter, but it would be nice to also have a standalone version again. Can you share the py2exe setup.py script that was used in the past?

**Comment by GammaC0de** [comment]
@treysis Windows package is now available.

**Comment by treysis** [comment]
Thx. However, the included pycurl/libcurl-library (pycurl 7.43.0, libcurl 7.47.0) doesn't work on dualstack-systems (can't resolve hostnames). It should be exchanged for the most recent pycurl 7.43.0.2, libcurl 7.60.0.

**Comment by GammaC0de** [comment]
pycurl 7.43.0 is the latest available pycurl  windows package - see https://dl.bintray.com/pycurl/pycurl/

**Comment by treysis** [comment]
Due to automatic sorting, unfortunately it's not at the end of the list, but a little bit further up, 7.43.0.2:
https://dl.bintray.com/pycurl/pycurl/:pycurl-7.43.0.2.win32-py2.7.zip

**Comment by GammaC0de** [comment]
Package updated

**Comment by obi12341** [comment]
Finally :)
When will we hit version 1.0.0? xD

**Comment by MaluNoPeleke** [comment]
What are the differences between pyplex and modern theme?

**Comment by GammaC0de** [comment]
> What are the differences between pyplex and modern theme?

Basically they are the same,
pyplex is the "dark" theme and modern is the "light" theme

**Comment by felixmosh** [comment]
I want to develop the webui with modern stack as spa, what will be the best way to run a dev env? I’m a Js dev with a minimal python knowledge.

**Comment by StefftheEmperor2** [comment]
I think we should develop a decoupled webui which interacts with pyload via the api

**Comment by felixmosh** [comment]
> I think we should develop a decoupled webui which interacts with pyload via the api

This is what I had in mind, pyload exposes an api & serves static files, the rest is done at client side.

The question is how I serve a Dev environment?

**Comment by treysis** [comment]
I think that'd be something for pyLoad-ng/0.5. I'd set zp dev env with pyCharm for pyLoad and then whatever suits you for what you want to use for the interface.

**Comment by felixmosh** [comment]
> I think that'd be something for pyLoad-ng/0.5. I'd set zp dev env with pyCharm for pyLoad and then whatever suits you for what you want to use for the interface.

Where Can I get the pyload-Ng?

**Comment by treysis** [comment]
I believe it's the master branch of this repo.

**Comment by julianrupp** [comment]
Using pyload in a docker container (writl/pyload) on Synology NAS.

Previously ran all fine, then recently I ran into issues when trying to use it with host rapidgator.net: Sometimes timeout errors when trying to add a premium account, or if I managed to add the account, frequently it would clear the password after a while, display as being invalid and fall back into free download mode. Two different hosters worked flawlessly, including premium account functionality.

So I went and updated to 0.4.20 (whole new docker image). Now, while Rapidgator works a little more reliably, I have noticed another new problem: a massive memory leak. Container booted up merely 40 minutes ago and already hogs more than 16 GB of RAM.

I have yet to test if this only occurs when downloading files from Rapidgator or if it also happens with other hosters, but at this rate it makes this new pyload version useless for longer unattended downloading. Since RG was the only hoster that I ran into issues with when using the 0.4.9 docker container, I'll likely revert to the old version until 0.4.20 is comparatively stable.

**Comment by GammaC0de** [comment]
@julianrupp Doker images are not official, they are maintained by third parties.
There is also https://hub.docker.com/r/linuxserver/pyload that you can try.

If you want, you can try to open a new issue regarding  the RG account,
There I can try to see if there is a problem with the account plugin.

**Comment by julianrupp** [comment]
Further testing --

With current setup, memory leak also exists when pyload is idle or extracting archives. Container process list however shows that the python process merely uses a bit over 300 MB of memory and unrar a negligible amount. No other processes in list.

So the memleak appears to be an issue with the container and not pyload.

I still frequently have issues that pyload will corrupt the saved account details for the RG premium account however, or even when the account info is still there, for unknown reasons new downloads on this host won't start but end up in a waiting loop. As a workaround, I frequently have to restart the container (when account info is corrupted) and/or cancel and restart indvidual files' downloads, sometimes multiple times in a row.

I've also tried the latest build of the other docker image, and this phenomenon also happens there. And it also has a memory leak. Container up for 14 minutes and already hogging over 8 GB. (Docker container process list shows python only using 50 MB and no other processes that eat all the RAM).

Is it possible that I/O behavior has changed in pyload since 0.4.9? Maybe log/debug output keeps getting written in some buffer now which the Docker container holds in RAM.

**Comment by GammaC0de** [comment]
First of all, your issues with RG/DOCKER is not related to this topic subject,
so please enable DEBUG mode and open a NEW issue for this providing logs, 

thanks.

**Comment by julianrupp** [comment]
Using the debug log, I traced the issues with Rapidgator back to occasional timeouts for API calls. Apparently the issue is on their side, not on pyLoad's. Looks like their servers are flaky.

After digging a while on the internet, it appears that I also found the cause of the memory leak and a fix for it. The container is now hovering between 900 MB and 1.75 GB RAM and does not grow further so far. If anyone having the same issue finds this page, here is my experimental solution: Modify the startup script inside the container so that STDOUT and STDERR of the python process are redirected to the null device. For the linuxserver/pyload container, the file you need to modify is /etc/services.d/pyload/run. Add a `>/dev/null 2>&1` at the end of the exec command. Also reported this to the devs of the container.

**Comment by obi12341** [comment]
why should stdout and stderr increase the memory consumption? This makes no sense...
